### PR TITLE
Added test case for `internal/errorgroup`

### PR DESCRIPTION
**Pull request by hlts2**
### Description:

I added a test case for `internal/errorgroup`.

### Environment:

- Go Version: 1.14.3
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.11.5

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/errgroup/group_test.go [hunk 1]
```diff
@@ -46,7 +46,8 @@ func TestNew(t *testing.T) {
 		afterFunc  func(args)
 	}
 	defaultCheckFunc := func(w want, got Group, got1 context.Context) error {
-		if !reflect.DeepEqual(got, w.want) {
+		if got, want := got.(*group), w.want.(*group); !reflect.DeepEqual(got.emap, want.emap) &&
+			!reflect.DeepEqual(got.enableLimitation, want.enableLimitation) {
 			return errors.Errorf("got = %v, want %v", got, w.want)
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:51]: do we need to check if `cancel` field is not nil?

File: internal/errgroup/group_test.go [hunk 2]
```diff
@@ -307,86 +328,78 @@ func Test_group_Limitation(t *testing.T) {
...
-		           emap: nil,
-		           errs: nil,
-		           err: nil,
-		           },
-		           want: want{},
-		           checkFunc: defaultCheckFunc,
-		       }
-		   }(),
-		*/
+		func() test {
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:357]: how about rewrite it like this? I think it is easier to read, what do you think?
```golang
357         {
358             name: "set disable when limit is 0",
359             args: args{
360                 limit: 0,
361             },
362             fields: fields{
363                 limitation:       make(chan struct{}),
364                 enableLimitation: atomic.Value{},
365             },
366             want: want{
367                 want: &group{
368                     enableLimitation: func() atomic.Value {
369                         var el atomic.Value
370                         el.Store(false)
371                         return el
372                     }(),
373                 },
374             },
375         },
```

File: internal/errgroup/group_test.go [hunk 3]
```diff
@@ -307,86 +328,78 @@ func Test_group_Limitation(t *testing.T) {
...
+				},
+				want: want{
+					want: &group{
+						enableLimitation: el,
+					},
+				},
+			}
+		}(),
+
+		func() test {
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:378]: same as before

File: internal/errgroup/group_test.go [hunk 4]
```diff
@@ -307,86 +328,78 @@ func Test_group_Limitation(t *testing.T) {
...
-		   }(),
-		*/
+		func() test {
+			var el atomic.Value
+			el.Store(false)
+
+			return test{
+				name: "set disable when limit is 0",
+				args: args{
+					limit: -1,
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:364]: it should be 0?

File: internal/errgroup/group_test.go [hunk 5]
```diff
@@ -511,102 +586,58 @@ func Test_group_Go(t *testing.T) {
 			if test.checkFunc == nil {
 				test.checkFunc = defaultCheckFunc
 			}
-			g := &group{
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:514]: I think it is better than `generator`, even it has lint error. what do you think?

Comment (kevindiu) [inline internal/errgroup/group_test.go:514]: I think it is better than `generator`, what do you think? (even it has lint error)

Comment (hlts2) [reply internal/errgroup/group_test.go:514]: I understand your opinion.
but I think we do not ignore lint error.
in this case, copying the mutex causes an error.
what do you think.

Comment (kevindiu) [reply internal/errgroup/group_test.go:514]: maybe just remove line 521?

File: internal/errgroup/group_test.go [hunk 6]
```diff
@@ -421,87 +426,157 @@ func Test_group_Go(t *testing.T) {
...
+
+			var enableLimitation atomic.Value
+			enableLimitation.Store(true)
+
+			egctx, cancel := context.WithCancel(context.Background())
+			cancel()
+
+			g := &group{
+				egctx:            egctx,
+				limitation:       make(chan struct{}),
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:455]: It should be 
```suggestion
				limitation:       make(chan struct{}, 0),
```

File: internal/errgroup/group_test.go [hunk 7]
```diff
@@ -421,87 +426,157 @@ func Test_group_Go(t *testing.T) {
...
+				enableLimitation: enableLimitation,
+			}
+
+			return test{
+				name: "f is not called when reached limit and cancel g.egctx",
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:460]: I think we also need to test the limitation function (for example the limit is 3, and 5 `Go(f func error)` is executed and only 3 can be executed at first, any after 1 is completed the 4th func will execute)

File: internal/errgroup/group_test.go [hunk 8]
```diff
@@ -421,87 +426,157 @@ func Test_group_Go(t *testing.T) {
...
+					cancel()
+					atomic.AddInt32(&cancelCalledCnt, 1)
+				},
+				emap:             make(map[string]struct{}),
+				limitation:       make(chan struct{}, 1),
+				enableLimitation: enableLimitation,
+			}
+
+			return test{
+				name: "f is called and f returns error",
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:547]: do we need to check multiple error is returned?

File: internal/errgroup/group_test.go (hunk 5 above)
Comment (hlts2) [reply internal/errgroup/group_test.go:514]: ok.
but if the test use mutex or sync.once variable, I use a generator.
what do you think?

File: internal/errgroup/group_test.go [hunk 9]
```diff
@@ -729,49 +738,34 @@ func Test_group_Wait(t *testing.T) {
...
-		           emap: nil,
-		           errs: nil,
-		           err: nil,
-		           },
-		           want: want{},
-		           checkFunc: defaultCheckFunc,
-		       }
-		   }(),
-		*/
+		{
```
Comment (kevindiu) [inline internal/errgroup/group_test.go:741]: do we need some test case to block the current `Go(func())` call and the `Wait()` should actually wait for all functions return?

File: internal/errgroup/group_test.go (hunk 6 above)
Comment (hlts2) [reply internal/errgroup/group_test.go:455]: I don't think you need to write 0 explicitly.

File: internal/errgroup/group_test.go (hunk 5 above)
Comment (kevindiu) [reply internal/errgroup/group_test.go:514]: using generator is the same. the difference of `generator` is that it is missing this statement `mu:  test.fields.mu,` , so the lint wouldn't report a error.

File: internal/errgroup/group_test.go (hunk 6 above)
Comment (kevindiu) [reply internal/errgroup/group_test.go:455]: One is Buffered channel and one is unbuffered channel, if you think using unbuffered one is okay then it is fine for me;
https://medium.com/a-journey-with-go/go-buffered-and-unbuffered-channels-29a107c00268

File: internal/errgroup/group_test.go (hunk 5 above)
Comment (hlts2) [reply internal/errgroup/group_test.go:514]: :memo:
we may use generator to avoid coping `waitGroup` & `sync.Mutex`

so I close this comment.:pray:

**Pull request closed without merging**